### PR TITLE
[AC-7777] Extend Office Hours calendar for finalists and staff

### PR DESCRIPTION
**Pull request by jonkiparsky**
https://masschallenge.atlassian.net/browse/AC-7777

Some things to check:
- appropriate responses for staff, finalists, and mentors
- users w/o office hours grants see no hours
in addition to office hours, we get a list of locations for which an office hours holder can hold office hours (locations associated with programs for which the target user holds an office-hour holding program role grant) and a list of startups associated with the user

### CODE REVIEW

File: web/impact/impact/v1/views/office_hours_calendar_view.py [hunk 1]
```diff
@@ -64,31 +78,107 @@ def _get_target_user(self, request):
...
+            reserved_by_user | unreserved & relevant_mentors,
+            start_date_time__range=[self.start_date, self.end_date])
+
+    def _null_office_hours_queryset(self):
+        return MentorProgramOfficeHour.objects.none()
+
+    def _get_office_hours_data(self):
+        primary_industry_key = "mentor__expertprofile__primary_industry__name"
+        office_hours = self._office_hours_queryset().filter(
+            program__isnull=True).order_by(
```
Comment (geofrocker) [inline web/impact/impact/v1/views/office_hours_calendar_view.py:174]: Was this intentional? this currently filters out all the office hours that are associated with a program.

**Review by geofrocker — CHANGES_REQUESTED** [review]
from QA
- A mentor gets a response with location_choices populated with null values
```json
"calendar_data": [],
    "location_choices": [
        {
            "location_id": null,
            "location_street_address": null,
            "location_timezone": null,
            "location_country": null,
            "location_state": null,
            "location_name": null,
            "location_city": null
        }
    ],
    "timezones": [],
```
  Shouldn't this be an empty array?
- As a staff user, I can get all office-hour sessions for all mentor associated with my programs ✅
- As a finalist, I can view office hours from relevant mentors ✅
- As a mentor, I can view my scheduled office hours ✅
@hfreund @jonkiparsky  Are we interested in office hours whose end_date is in the past for the frontend?

File: web/impact/impact/v1/views/office_hours_calendar_view.py (hunk 1 above)
Comment (hfreund) [reply web/impact/impact/v1/views/office_hours_calendar_view.py:174]: Yes, I asked Jon to do this so office hours created on the old (current) calendar do not appear on the new calendar. They have different durations which causes some issues that didn't seem worth resolving - it seemed easier to maintain separation between the two interfaces.

**Comment by hfreund** [comment]
> Are we interested in office hours whose end_date is in the past for the frontend?

@geofrocker yes we would like to be able to display past office hours in the front-end.

**Review by iankigen — APPROVED** [review]
Code looks good

File: web/impact/impact/v1/views/office_hours_calendar_view.py [hunk 2]
```diff
@@ -118,50 +208,83 @@ def _get_office_hours_data(self):
         self.response_elements['location_choices'] = self.location_choices()
         program_families = self.mentor_program_families()
         self.response_elements['mentor_program_families'] = program_families
+        self.response_elements['user_startups'] = self._user_startups()
         self.succeed()
 
+    def _user_startups(self):
+        return self.target_user.startupteammember_set.order_by(
+            '-id').values(
+                "id",
```
Comment (iankigen) [inline web/impact/impact/v1/views/office_hours_calendar_view.py:217]: `id` should be the `startup__id` and this list should also be `distinct()`

Comment (iankigen) [inline web/impact/impact/v1/views/office_hours_calendar_view.py]: from QA:
`id` should be the `startup__id` and this list should also be `distinct()`

**Review by iankigen — CHANGES_REQUESTED** [review]

**Review by geofrocker — APPROVED** [review]
looks good ✅

File: web/impact/impact/v1/views/office_hours_calendar_view.py (hunk 2 above)
Comment (jonkiparsky) [reply web/impact/impact/v1/views/office_hours_calendar_view.py]: good catch! Fixed.

**Review by iankigen — APPROVED** [review]

**Pull request merged**